### PR TITLE
Merge data-products branch into main

### PR DESCRIPTION
**Pull request by bocchino**
Now that FPP v2.0.0 is released, we can merge the ongoing data products work into the main branch.

### CODE REVIEW

**Comment by bocchino** [comment]
@LeStarch The auto-tracing may not be working as expected. The build failed until I updated the trace manually. See https://github.com/fprime-community/fpp/pull/341/commits/48ed114191682cc4ef1e890ac72a8ce7a54baf82.

File: compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala [hunk 1]
```diff
@@ -432,25 +472,133 @@ case class ComponentTesterBaseWriter(
...
+      lines(
+        """|DpSend e = { id, buffer };
+           |this->productSendHistory->push_back(e);"""
+      )
+    )
+    lazy val handleProductSend = functionClassMember(
+      Some(
+        """|Handle a data product send from the component under test
+           |
+           |By default, call pushProductRequestEntry. You can override
```
Comment (tiffany1618) [inline compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala:558]: `pushProductRequestEntry` -> `pushProductSendEntry`?

File: compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala [hunk 2]
```diff
@@ -432,25 +472,133 @@ case class ComponentTesterBaseWriter(
...
-              |"""
+          s"""|FW_ASSERT(this->$portNumGetter() > 0);
+              |FW_ASSERT(this->$varName[0].isConnected());
+              |this->$varName[0].invoke(id, buffer, status);"""
         )
       }
-    ).getOrElse(Nil)
+    )
+    lazy val pushProductSend = functionClassMember(
+      Some("Push an entry on the product request history"),
```
Comment (tiffany1618) [inline compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala:545]: request -> send?

**Review by tiffany1618 — COMMENTED** [review]
Looks great! Just a few comments.

File: compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala (hunk 1 above)
Comment (bocchino) [reply compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala:558]: Good catch!

File: compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala (hunk 2 above)
Comment (bocchino) [reply compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/TestWriter/ComponentTesterBaseWriter.scala:545]: Good catch!

File: compiler/tools/fpp-to-cpp/test/component/base/ActiveAsyncProductsComponentAc.ref.hpp [hunk 3]
```diff
@@ -0,0 +1,1665 @@
...
+    //! The container ids
+    struct ContainerId {
+      enum T : FwDpIdType {
+        Container1 = 100,
+        Container2 = 200,
+        Container3 = 300,
+        Container4 = 400,
+        Container5 = 500,
+      };
+    };
```
[the remark this answers is not in the document]
Comment (bocchino) [reply compiler/tools/fpp-to-cpp/test/component/base/ActiveAsyncProductsComponentAc.ref.hpp:117]: I like the idea of using an enum class. There is still some concern about using C++11 features that aren't easily translatable to C++98 (e.g., by deleting the override keyword). I propose that we merge the PR as is and hold this as an open issue.

[the remark this answers is not in the document]
Comment (bocchino) [reply compiler/tools/fpp-to-cpp/test/component/base/ActiveAsyncProductsComponentAc.ref.hpp:117]: @LeStarch Your thoughts?

[the remark this answers is not in the document]
Comment (LeStarch) [reply compiler/tools/fpp-to-cpp/test/component/base/ActiveAsyncProductsComponentAc.ref.hpp:117]: Agreed.  Track as an issue and we will resolve.

**Review by LeStarch — APPROVED** [review]

**Pull request merged**